### PR TITLE
Calypso Framework: Request tracking + deduping middleware

### PR DESCRIPTION
**Pull request by samouri**
Within the Calypso world, the current status quo is to track requests using the redux state tree and reducers.  This leads to bloat (essentially the same request tracking code being written over and over) as well as performance degradations (the less in the state tree the better).

This PR presents a proposal for an alternative where we can track inflight requests using custom middleware and [FSA](https://github.com/acdlite/flux-standard-action) compliant actions.

actions would specify either key `requestStart` or `requestEnd` in their meta.  The value would be equal to the dedupe key.  e.g.

```jsx
export function requestTeams() {
	return {
		type: READER_TEAMS_REQUEST,
		meta: {
			requestStart: READER_TEAMS_REQUEST
		}
	};
}

export function receiveTeams( { payload, error } ) {
	return {
		type: READER_TEAMS_RECEIVE,
		payload,
		error: !! error,
		meta: {
			requestEnd: READER_TEAMS_REQUEST,
		}
	};
}
```

A more fun example utilizing a request with params could be:
```jsx
export function requestSite( siteId ) {
	return {
		type: READER_SITE_REQUEST,
		meta: {
			requestStart: `${READER_SITE_REQUEST}:${siteId}`: 
		}
	};
}

export function receiveSite( { payload, error, siteId } ) {
	return {
		type: READER_SITE_RECEIVE,
		payload,
		error: !! error,
		meta: {
			requestEnd: `${READER_SITE_REQUEST}:${siteId}`,
		}
	};
}
```

When these actions are dispatched the middleware would pick these up and either eat the action if it is already inflight or allow it through based on their request keys.  No more storing this data in reducers or writing selectors for things like `isRequestingTeams`.  Query components also become much simpler!

#### FAQ:
**Q**: won't the http middleware obviate the need for this?
**A**: while http middleware is a better alternative for many cases, this deduping library is more flexible in that it does not rely on http being used and/or the exact url being accessed.

Note: included in this PR is a sample usage with `ReaderTeams` component. It is a prime target for this middleware because we do not care about showing any sort of progress indicator as it is loading .

### CODE REVIEW

File: client/state/reader/teams/actions.js [hunk 1]
```diff
@@ -7,11 +7,26 @@
...
 } from 'state/action-types';
 
 export function requestTeams() {
 	return {
 		type: READER_TEAMS_REQUEST,
+		meta: {
+			requestStart: READER_TEAMS_REQUEST
+		}
+	};
+}
```
Comment (gwwar) [inline client/state/reader/teams/actions.js]: Is there anything we can do here via convention or composition to make adding the dedupe meta easier?

Comment (samouri) [inline client/state/reader/teams/actions.js]: Definitely! 
I think it may be helpful to add some helpers for this but I rather that be in a different pr

Comment (samouri) [inline client/state/reader/teams/actions.js]: @gwwar do you have any gut guesses here?  I had a couple ideas:
1. helper func possibility: serialize the whole action and use it as the requestKey 
2. potential convention: `${action_type}:${identifier}`

Comment (gwwar) [inline client/state/reader/teams/actions.js]: > helper func possibility: serialize the whole action

One thing that comes to mind is that both the keys for request and receive actions need to match. Most receive actions just have the type and item, so adding any request params onto the signature would be a change.

My intuition is that we should keep the actions rather simple and straightforward, while we wrap the meta at a higher layer. The data layer can maybe pass through meta props we're still interested in.

A quick thought:
```javascript
// maybe in the query component?
requestStart( key, requestAction() );
//in the data-layer?
const dataHandler = ( { dispatch }, action ) => {
    const key = get( action, 'meta.requestStart' );
    //...
   return promise.then( () => {
        dispatch( mySuccessAction() );
        dispatch( requestEnd( key, myReceiveAction() ) );
    } )
    .catch( () => {
        dispatch( requestEnd( key, myErrorAction() ) );
    } );
}
```

Comment (blowery) [inline client/state/reader/teams/actions.js]: The thing that bugs me a bit here is that we're leaking knowledge. I'm not sure who really needs to know that we need to dedupe requests, outside of the data middleware. It's not a terrible leakage, as the requestKey could be used for other things I suppose. 

I'm not sure I love `requestStart` as the kickoff. Yes, I know I suggested it. :) 

From the action creator's standpoint, something like `keyedRequest( baseAction )` would probably make more sense? Then the data-layer would dispatch something like `completeKeyedRequest( baseAction )`

Comment (blowery) [inline client/state/reader/teams/actions.js]: then `keyedRequest` aka `requestStart` could also take a second param, which is the actual value of the key. The default behavior would be to deterministically serialize the action.

Comment (gwwar) [inline client/state/reader/teams/actions.js]: @blowery I think you're right here about the separation of concerns.

Let me catch up on what's going on in #10327, if it does provide us with a `pure and stateless descriptions of those desired HTTP requests` we should be able to implement a simple dedupe strategy on top of it without leakage.

Comment (blowery) [inline client/state/reader/teams/actions.js]: @gwwar mostly, yeah. unfortunately I think I have a case where a pure URL dedupe isn't necessarily enough.

When we ask for updates to a stream, we ask for a after our most recent post and after the client's Date.now(). Since we use the client's `now`, the URL changes every time our update poller triggers. I think we're mostly safe from dedupes here for other reasons, but something to consider.

**Comment by dmsnell** [comment]
Things I like about this:
 - It's standard and FSA and all that
 - Requests get a unique key (cross-reference with @coderkevin)
 - Auto dedupe

Things I don't like about this:
 - It continues to encourage strong coupling between our reducers and the API responses
 - `_SUCCESS`, `_FAILURE`, and `_RECEIVE` actions which pass opaque data structures (makes it hard to inject actions from the console dispatcher or over websockets where the data mechanisms don't share the same idioms as HTTP requests use

Ultimately I think it's a solution tied to the idea that we are abstracting over the API and building our app thinking that it is displaying remote data.

The approach I have taken is that we are abstracting over types of data objects and we have the source of the truth in the running Calypso.

For comparison:
 - HTTP request deduce is/almost is trivial once we get that HTTP middleware in. HTTP is a language in and of itself
 - each endpoint handlers in the current API middleware can use (and I hope they do) the keyed approach to deduce. in my mind this is as simple as using a mutable `Map()` where it simply checks `if ( requestsInFlight.has( `${ siteId }-${ postId }` ) ) { … }` blah blah blah. easy peasy.

Sorry this is terse. I was out all day and trying to pay catch-up

**Comment by samouri** [comment]
> It continues to encourage strong coupling between our reducers and the API responses

not sure what you mean by this.  The api-middleware can perform any transformations it want to api responses. `payload` does _not_ mean `api response dump`.  It is just a standard way of structuring actions.

> _SUCCESS, _FAILURE, and _RECEIVE actions which pass opaque data structures (makes it hard to inject actions from the console dispatcher or over websockets where the data mechanisms don't share the same idioms as HTTP requests use

1. _SUCCESS & _FAILURE are already out.  
2. By opaque data structures I think you are talking about the same issue highlighted in the previous point.  That there should be a transformation from the api response to a format that Calypso cares more about (and is hopefully explicitly verified in the data-layer). This concern is also completely separate to deduping requests by arbitrary key.

> For comparison:

I _do_ think HTTP deduping is going to be significantly better and come for free in the vast majority of cases.  The point of this PR is twofold:

1. allow for deduping on arbitrary key 
1. alleviate the need for a deduping now ( so we don't need to keep creating.

The criticism that I most feel agree with so far is the one that @blowery pointed out re. separation of concerns:
> The thing that bugs me a bit here is that we're leaking knowledge. I'm not sure who really needs to know that we need to dedupe requests, outside of the data middleware. 

It convinced me that this is the wrong abstraction for now and I'd rather go with dumbcode: https://github.com/Automattic/wp-calypso/pull/11069

**Comment by samouri** [comment]
closing in favor of #11069
